### DISCUSSION TITLE
TanStack Router with Clerk

### DISCUSSION THREAD

**Discussion by JORDAAAN1495** [TanStack/router#1119] [Q&A]
I am trying to test out TanStack Router with Clerk and facing some difficulties with the auth and using the the clerk or convex hooks. Has anyone had any luck at setting up auth? I find that I get stuck in an infinite loop of continuously logging in. I have seen a similar post regarding clerk on here and tried that implementation, however no luck.

This is using the latest packages as of today (brand new app via Vite).

This is my `__root.tsx`
```typescript
import type { LoadedClerk } from "@clerk/types";
import { Outlet, createRootRouteWithContext } from "@tanstack/react-router";

interface MyRouterContext {
  isSignedIn: boolean;
  auth?: LoadedClerk;
}

export const Route = createRootRouteWithContext<MyRouterContext>()({
  component: () => <Outlet />,
});
```
This is the `index.tsx`
```typescript
import { useUser } from "@clerk/clerk-react";
import { createFileRoute } from "@tanstack/react-router";

export const Route = createFileRoute("/")({
  beforeLoad: async ({ context }) => {
    if (!context.isSignedIn) {
      await context.auth?.redirectToSignIn();
    }
  },
  component: Index,
});

function Index() {
  const { user } = useUser();

  return <>{user?.fullName}</>;
}
```
This is the `App.tsx`
```typescript
import { ClerkProvider, useAuth, useClerk } from "@clerk/clerk-react";
import { RouterProvider } from "@tanstack/react-router";
import { ConvexReactClient } from "convex/react";
import { ConvexProviderWithClerk } from "convex/react-clerk";
import router from "./router";
import appSettings from "./utils/appSettings";

function InnerApp() {
  const { isSignedIn } = useAuth();
  const auth = useClerk();

  return <RouterProvider router={router} context={{ isSignedIn, auth }} />;
}

function App() {
  const convex = new ConvexReactClient(appSettings.convex.url);

  return (
    <ClerkProvider publishableKey={appSettings.clerk.publishableKey}>
      <ConvexProviderWithClerk client={convex} useAuth={useAuth}>
        <InnerApp />
      </ConvexProviderWithClerk>
    </ClerkProvider>
  );
}

export default App;
```
And finally this is the `router.tsx`
```typescript
import { createRouter } from "@tanstack/react-router";
import { routeTree } from "./routeTree.gen";

const router = createRouter({
  routeTree,
  context: {
    isSignedIn: false,
    auth: undefined,
  },
});

declare module "@tanstack/react-router" {
  interface Register {
    router: typeof router;
  }
}

export default router;
```

**Comment by keduong33** [comment]
Have you figured out the solution?

**Comment by ssijak** [comment]
https://tanstack.com/router/v1/docs/framework/react/examples/start-clerk-basic

**Reply by chinanderm** [reply]
This is for Start, not Router.

**Reply by developerEhsan** [reply]
@chinanderm I explained the solution here -> https://medium.com/@your-ehsan/how-to-setup-clerk-with-tanstack-router-in-vite-react-with-proper-type-safety-using-typescript-80af12a09771

**Comment by aldotestino** [comment]
I found two possible solutions:

**1. Use `beforeLoad` to check for the token and redirect:**

This approach checks the token during the route loading phase and redirects the user if the token is missing.

```typescript
import { createFileRoute, Outlet, redirect } from '@tanstack/react-router';

export const Route = createFileRoute('/_authed')({
  beforeLoad: async ({ context }) => {
    // Fetch the token before rendering the route
    const token = await context.auth.getToken();
    // Redirect to sign-in if token is missing
    if (!token) {
      throw redirect({
        to: '/sign-in',
      });
    }
  },
  component: Outlet,
});
```

**2. Use Clerk's `<SignedIn>` and `<SignedOut>` components:**

This approach is more declarative, using Clerk's built-in components to conditionally render based on the user's signed-in state.

```typescript
import { SignedIn, SignedOut } from '@clerk/clerk-react';
import { createFileRoute, Navigate, Outlet } from '@tanstack/react-router';

export const Route = createFileRoute('/_authed')({
  component: AuthedLayout,
});

function AuthedLayout() {
  return (
    <>
      {/* Render Outlet if user is signed in */}
      <SignedIn>
        <Outlet />
      </SignedIn>
      {/* Redirect to sign-in if user is signed out */}
      <SignedOut>
        <Navigate to="/sign-in" />
      </SignedOut>
    </>
  );
}
```

**Reply by mikr13** [reply]
How are you getting `.getToken()` in context.auth? `LoadedClerk` doesn't have `getToken` method as far as I can see. It has session & other stuff!

**Reply by aldotestino** [reply]
Sorry, my `auth` is the return of `useAuth()` and not of `useClerk()`

**Reply by mikr13** [reply]
Makes sense. Thanks!

**Reply by chinanderm** [reply]
@aldotestino Are you not running into issues with the result of `useAuth()` being a new object reference? For me, it changes every time Clerk's touchSession stuff runs.

Edit: Nevermind! I'm reading the now-present [docs](https://tanstack.com/router/v1/docs/framework/react/guide/router-context#how-about-using-react-contexthooks) on providing context to the router via <RouterProvider /> instead of baking into the router object.

**Reply by developerEhsan** [reply]
For anyone coming to this thread see for solution -> https://medium.com/@your-ehsan/how-to-setup-clerk-with-tanstack-router-in-vite-react-with-proper-type-safety-using-typescript-80af12a09771

**Reply by kein-1** [reply]
docs have a good example:

https://tanstack.com/router/v1/docs/framework/react/how-to/setup-auth-providers#2-set-up-environment-variables-1

basically in your option 1: you'll just expose a boolean flag from the useAuth context and check

Im running into issues with flickering though; wondering if anyone has a solution:

```
const App = () => {
  const auth = useClerkAuth()

  if (auth.isLoading) {
    return (
      <div className="flex min-h-screen items-center justify-center">
        Loading... // this part flickers on page reloads
      </div>
    )
  }

  return (
    <QueryClientProvider client={queryClient}>
      <RouterProvider router={router} context={{ auth }} />
    </QueryClientProvider>
  )
}```

**Comment by mhuggins** [comment]
The problem with this approach is that it requires `RouterProvider` to be wrapped inside of `ClerkProvider`.  However, I want to be able to define `routerPush` and `routerReplace` props on `ClerkProvider`, which necessitates it being wrapped inside `RouterProvider`.  

I'm currently doing this:

```ts
// routes/__root.tsx
const Route = createRootRoute({
  component: Root,
});

function Root() {
  const navigate = useNavigate();

  return (
    <ClerkProvider
      publishableKey={clerkPublishableKey}
      routerPush={(to) => navigate({ to })}
      routerReplace={(to) => navigate({ to, replace: true })}
    >
      <Outlet />
    </ClerkProvider>
  );
};

// main.tsx
root.render(<RouterProvider router={router} />);
```

Because I need access to `useNavigate` to specify those router-related props, I cannot reverse the order of these two components.  Maybe there's a setter that can be used to update the provider after it's already in the page?

**Reply by Fatih-Ertikin** [reply]
You don't necessarily have to wrap `ClerkProvider` inside `RouterProvider`: You can use `router.navigate` to navigate outside of react (even when `RouterProvider` is lower down in the JSX tree).

I think this also fixes @kein-1 problem

```jsx
// main.tsx

function InnerApp() {
  const auth = useAuth();

    // Should only trigger once on app mount
    if (!auth.isLoaded) {
    return <div style={{ backgroundColor: "red", height: "100vh" }}>some loader</div>;
  }

  return <RouterProvider router={router} context={{ auth }} />
}


createRoot(document.getElementById("root")!).render(
  <StrictMode>
    <ClerkProvider
      publishableKey={import.meta.env.VITE_CLERK_PUBLISHABLE_KEY}
      routerPush={(to) => router.navigate({ to, replace: false })} // "router" object comes from createRouter (in main.tsx) left out for clarity
      routerReplace={(to) => router.navigate({ to, replace: true })}
    >
      <InnerApp />
    </ClerkProvider>
  </StrictMode>,
);
```
